### DISCUSSION TITLE
Write an object with Show-WithOutEmpty

### DISCUSSION THREAD

**Discussion by jdhitsolutions** [SamErde/PSPreworkout#17] [General]
Instead of writing a hashtable with `Show-WithoutEmpty`,  I suggest writing a custom object.

```powershell
$g.psobject.properties | where {$_.Value} | ForEach-Object -begin {$h = @{}} -Process {$h.Add($_.name,$_.Value)} -End {New-Object -Typename PSObject -Property $h}
```

This makes it easier to format or work with the output.  I also think the command should take pipeline input so I can run a command like this:

```powershell
Get-CimInstance win32_operatingsystem | Show-WithoutEmpty | Out-GridView
```

**Comment by SamErde** [comment]
I'm struggling to make your example work. I'm probably misunderstanding something in it, but finally got this version to work after wandering through the zone of trial and error:

```powershell
    [PSCustomObject]$Object = @{
        Name          = 'Sam'
        Colors        = @('blue', 'green')
        FavoriteError = $null
        DeskHeight    = $null
        DeskWidth     = $null
        ChairComfort  = 5
    }
    [PSCustomObject]$PropertiesWithValues = @{}
    [array]($Object.psobject.properties.Item('Keys').Value) | ForEach-Object {
        if (($Object.psobject.Properties.Item('SyncRoot').Value)[$_]) {
            $PropertiesWithValues.Add($_, ($Object.psobject.Properties.Item('SyncRoot').Value)[$_])
        }
    }
    $PropertiesWithValues
```
☝️ This can't be the best way!

Otherwise, it gives me this:
![image](https://github.com/user-attachments/assets/acd4ea73-e80a-4877-9006-beba4d0f1f5d)

**Comment by jdhitsolutions** [comment]
Part of your problem is that you are not creating a custom object. This:

```powershell
 [PSCustomObject]$Object = @{
```
is creating a hashtable. You need this:

```powershell
$Object = [PSCustomObject]@{
```

But this will break the rest of your code. So use mine:

```powershell
$Object.PSObject.properties | 
where {$_.Value} | 
ForEach-Object -begin {
    $h = [ordered]@{}
} -Process {
    $h.Add($_.name,$_.Value)
} -End {
    New-Object -Typename PSObject -Property $h
}
```

![image](https://github.com/user-attachments/assets/5148f527-2694-4e70-ad97-c673f5ae8e38)

**Comment by SamErde** [comment]
Oh, good grief. So, even when I was testing with this: 
```powershell
$Object = [PSCustomObject]::new(@{
    Name          = 'Sam'
    Colors        = @('blue', 'green')
    FavoriteError = $null
    DeskHeight    = $null
    DeskWidth     = $null
    ChairComfort  = 5
})
```
...my 'PSCustomObject' was *still* creating a hash table! (...and hash tables specifically do not work with this function for reasons that I was starting to suspect but have not *yet* fully grasped.)

Your code still does not work with my hashtable object, but if I save "any" other type of object to a variable, it DOES work. For example:

```powershell
$Object = Get-ChildItem .\PSPreworkout.psd1
$Object.PSObject.properties | 
    Where-Object { $_.Value } | 
        ForEach-Object -Begin {
            $h = [ordered]@{}
        } -Process {
            $h.Add($_.name, $_.Value)
        } -End {
            New-Object -TypeName PSObject -Property $h
        }
```

**Comment by SamErde** [comment]
I'm caught up now. I needed to convert my hash table into a PSObject.

```powershell
    [PSCustomObject]$Stuff = @{
        Name = 'Sam'
        Colors = @('blue', 'green')
        FavoriteError = $null
        DeskHeight = $null
        DeskWidth = $null
        ChairComfort = 5
    }
    $Object = New-Object -TypeName PSObject -Property $Stuff
    # or
    $Object = Get-ChildItem .\PSPreworkout.psd1
```

Now it all works. Thanks again!

**Comment by jdhitsolutions** [comment]
The `New()` method was still creating a hashtable. And you are still using `[PSCustomObject]` on the wrong side. 

```powershell
$Stuff = [PSCustomObject]@{
        Name = 'Sam'
        Colors = @('blue', 'green')
        FavoriteError = $null
        DeskHeight = $null
        DeskWidth = $null
        ChairComfort = 5
    }
$Stuff.PSObject.properties | 
    Where-Object { $_.Value } | 
    ForEach-Object -Begin {
        $h = [ordered]@{}
    } -Process {
        $h.Add($_.name, $_.Value)
    } -End {
        New-Object -TypeName PSObject -Property $h
    }
```

![image](https://github.com/user-attachments/assets/b24c14bd-4527-4023-832d-a407363c6b90)

Here's the difference.

![image](https://github.com/user-attachments/assets/8d2a0ed4-db8e-48ae-acda-06a021499ec6)

`[PSCustomObject]` doesn't behave like a normal type accelerator. You need to attach it directly to the hashtable.

**Reply by SamErde** [reply]
Ah, the example in my CBH? Thank you. Fixed and pushed [v1.4.3](https://github.com/SamErde/PSPreworkout/blob/main/src/PSPreworkout/Public/Show-WithoutEmptyProperty.ps1). 🙂
